File: src/simulation/dynamics/stateArchitecture/_UnitTest/test_effectorNamingIntegration.py
# Copyright (c) 2026, Autonomous Vehicle Systems Lab, University of Colorado at Boulder
# This file is distributed under the ISC License in LICENSE.

"""Exercise naming preparation across spacecraft, attachment trees, and shared managers."""

from dataclasses import dataclass

import numpy as np
import pytest

from Basilisk.architecture import bskLogging
from Basilisk.simulation import (
    constraintDynamicEffector,
    hingedRigidBodyStateEffector,
    prescribedMotionStateEffector,
    spacecraft,
    spacecraftSystem,
    spinningBodyOneDOFStateEffector,
    thrusterDynamicEffector,
)
from Basilisk.utilities import SimulationBaseClass, deprecated, macros


@dataclass
class Tree:
    """Retain the Python owners of every node until simulation workers stop."""

    vehicle: object
    roots: list
    leaves: list
    loads: list

    def state_names(self):
        """Return state names in physical attachment order."""
        return [name for root, leaf in zip(self.roots, self.leaves) for name in (
            root.nameOfsigma_PMState, leaf.nameOfThetaState, leaf.nameOfThetaDotState
        )]

    def values(self):
        """Read integrated states without depending on alphabetical map order."""
        return np.concatenate([
            np.asarray(self.vehicle.dynManager.getStateObject(name).getState()).ravel()
            for name in self.state_names()
        ])


def make_tree(local):
    """Construct in reverse order and reserve automatic candidates from a later branch."""
    leaves = [spinningBodyOneDOFStateEffector.SpinningBodyOneDOFStateEffector() for _ in range(2)][::-1]
    roots = [prescribedMotionStateEffector.PrescribedMotionStateEffector() for _ in range(2)][::-1]
    loads = [thrusterDynamicEffector.ThrusterDynamicEffector() for _ in range(2)]
    vehicle = spacecraft.Spacecraft()
    vehicle.hub.mHub = 100.0  # [kg]
    vehicle.hub.IHubPntBc_B = 100.0 * np.eye(3)  # [kg m^2]
    if local:
        import effectorNamingTestSupport

        effectorNamingTestSupport.enableManagerLocalNaming(vehicle.dynManager)
        leaves[1].nameOfThetaState = "prescribedObjectspinningBodyTheta1"
        leaves[1].nameOfInertialPositionProperty = "spinningBodyInertialPosition1"
        roots[1].nameOfPrescribedPositionProperty = "prescribedObjectPosition1"
    for index, (root, leaf, load) in enumerate(zip(roots, leaves, loads)):
        root.setR_PM_M([0.3 * (index + 1), 0.0, 0.0])  # [m]
        leaf.mass = 1.0  # [kg]
        leaf.sHat_S = [1.0, 0.0, 0.0]
        leaf.thetaInit = 0.125 * (index + 1)  # [rad]
        leaf.thetaDotInit = 0.02  # [rad/s]
        leaf.addDynamicEffector(load)
        root.addStateEffector(leaf)
        vehicle.addStateEffector(root)
    return Tree(vehicle, roots, leaves, loads)


@pytest.mark.parametrize("reverse_schedule", [False, True])
def test_independent_spacecraft_repeat_preparation_in_one_simulation(manager_local, reverse_schedule):
    """Independent managers isolate names and bindings through repeated initialization."""
    sim = SimulationBaseClass.SimBaseClass()
    process = sim.CreateNewProcess("process")
    step = 0.01  # [s]
    process.addTask(sim.CreateNewTask("task", macros.sec2nano(step)))
    trees = [make_tree(manager_local) for _ in range(2)][::-1]
    schedule = trees[::-1] if reverse_schedule else trees
    for tree in schedule:
        sim.AddModelToTask("task", tree.vehicle)
        for model in tree.roots + tree.leaves + tree.loads:
            sim.AddModelToTask("task", model)
    sim.InitializeSimulation()
    names = [tree.state_names() for tree in trees]
    initial = [tree.values() for tree in trees]
    if manager_local:
        assert names[0] == names[1]
        for tree in trees:
            assert tree.leaves[0].nameOfThetaState == "prescribedObjectspinningBodyTheta2"
            assert tree.leaves[1].nameOfThetaState == "prescribedObjectspinningBodyTheta1"
            assert tree.roots[0].nameOfPrescribedPositionProperty == "prescribedObjectPosition2"
    sim.InitializeSimulation()
    for tree, original_names, original_values in zip(trees, names, initial):
        assert tree.state_names() == original_names
        np.testing.assert_array_equal(tree.values(), original_values)
    # Identical property names in independent managers must still address different storage.
    first_name = trees[0].loads[0].getPropName_inertialPosition()
    second_name = trees[1].loads[0].getPropName_inertialPosition()
    first_manager = trees[0].vehicle.dynManager
    second_manager = trees[1].vehicle.dynManager
    first_original = np.array(first_manager.getPropertyReference(first_name))
    second_original = np.array(second_manager.getPropertyReference(second_name))
    marker = np.array([[3.0], [4.0], [5.0]])  # [m]
    first_manager.setPropertyValue(first_name, marker)
    np.testing.assert_array_equal(trees[0].loads[0].inertialPositionProperty, marker)
    np.testing.assert_array_equal(trees[1].loads[0].inertialPositionProperty, second_original)
    first_manager.setPropertyValue(first_name, first_original)
    stop = 0.02  # [s]
    sim.ConfigureStopTime(macros.sec2nano(stop))
    sim.ExecuteSimulation()
    np.testing.assert_allclose(trees[0].values(), trees[1].values(), rtol=0.0, atol=1e-14)
    assert np.all(np.isfinite(trees[0].values()))


@pytest.mark.parametrize("reverse_attachment", [False, True])
@pytest.mark.parametrize("reverse_initialization", [False, True])
def test_constraint_keeps_equal_names_in_independent_managers_distinct(
    naming_support, reverse_attachment, reverse_initialization
):
    """A cross-spacecraft constraint binds each parent's own storage in either order."""
    trees = [make_tree(True) for _ in range(2)]
    constraint = constraintDynamicEffector.ConstraintDynamicEffector()
    constraint.setAlpha(1.0)  # [1/s]
    constraint.setBeta(1.0)  # [1/s]
    parents = trees[::-1] if reverse_attachment else trees
    for tree in parents:
        tree.leaves[0].addDynamicEffector(constraint)
    initialization_order = trees[::-1] if reverse_initialization else trees
    for _ in range(2):
        for tree in initialization_order:
            tree.vehicle.initializeDynamics()
    assert len(constraint.getPropName_inertialPosition()) == 2
    assert len(set(constraint.getPropName_inertialPosition())) == 1
    for index, tree in enumerate(trees):
        manager = tree.vehicle.dynManager
        leaf = tree.leaves[0]
        position = np.array([[4.0 * index], [0.0], [0.0]])  # [m]
        manager.setPropertyValue(leaf.nameOfInertialPositionProperty, position)
        for part in ("Velocity", "Attitude", "AngVelocity"):
            manager.setPropertyValue(getattr(leaf, f"nameOfInertial{part}Property"), np.zeros((3, 1)))
    constraint.computeForceTorque(0.0, 0.01)  # [s]
    constraint.computeForceTorque(0.0, 0.01)  # [s]
    np.testing.assert_allclose(np.abs(constraint.forceExternal_N), [[4.0], [0.0], [0.0]], atol=1e-14)  # [N]


def make_legacy_system():
    """Build docked and undocked units sharing the deprecated system's manager."""
    with pytest.warns((deprecated.BSKDeprecationWarning, deprecated.BSKUrgentDeprecationWarning)):
        system = spacecraftSystem.SpacecraftSystem()
    free = spacecraftSystem.SpacecraftUnit()
    docked = spacecraftSystem.SpacecraftUnit()
    units = [system.primaryCentralSpacecraft, docked, free]
    panels = [hingedRigidBodyStateEffector.HingedRigidBodyStateEffector() for _ in units][::-1]
    loads = [thrusterDynamicEffector.ThrusterDynamicEffector() for _ in units]
    for index, (unit, panel, load) in enumerate(zip(units, panels, loads)):
        unit.spacecraftName = f"vehicle{index}"
        unit.hub.mHub = 100.0  # [kg]
        unit.hub.IHubPntBc_B = 100.0 * np.eye(3)  # [kg m^2]
        panel.mass = 1.0  # [kg]
        panel.IPntS_S = np.eye(3)  # [kg m^2]
        panel.thetaInit = 0.125 * (index + 1)  # [rad]
        panel.nameOfThetaState = "customAngle"
        panel.addDynamicEffector(load)
        unit.addStateEffector(panel)
    ports = [spacecraftSystem.DockingData(), spacecraftSystem.DockingData()]
    for unit, port, name in zip(units, ports, ("primaryPort", "dockedPort")):
        port.portName = name
        unit.addDockingPort(port)
    system.attachSpacecraftToPrimary(docked, "dockedPort", "primaryPort")
    system.addSpacecraftUndocked(free)
    return system, units, panels, loads, ports


def test_deprecated_system_preserves_legacy_names_in_its_shared_manager():
    """All units keep distinct legacy names and bindings through repeated initialization."""
    system, units, panels, loads, ports = make_legacy_system()
    system.initializeDynamics()
    names = [panel.nameOfThetaState for panel in panels]
    hub_names = [unit.hub.nameOfHubPosition for unit in units]
    property_names = [load.getPropName_inertialPosition() for load in loads]
    assert names == [f"vehicle{index}customAngle" for index in range(3)]
    assert len(set(load.getPropName_inertialPosition() for load in loads)) == 3
    system.initializeDynamics()
    assert names == [panel.nameOfThetaState for panel in panels]
    assert hub_names == [unit.hub.nameOfHubPosition for unit in units]
    assert property_names == [load.getPropName_inertialPosition() for load in loads]
    for panel, name in zip(panels, names):
        assert system.dynManager.getStateObject(name).getState()[0][0] == panel.thetaInit
    for index, load in enumerate(loads):
        marker = np.array([[float(index)], [2.0], [3.0]])  # [m]
        system.dynManager.setPropertyValue(load.getPropName_inertialPosition(), marker)
        np.testing.assert_array_equal(load.inertialPositionProperty, marker)


def test_deprecated_unit_rejects_new_policy_before_registering(naming_support):
    """Direct initialization cannot bypass the deprecated system's legacy-only boundary."""
    system, units, panels, loads, ports = make_legacy_system()
    naming_support.enableManagerLocalNaming(system.dynManager)
    with pytest.raises(bskLogging.BasiliskError, match="legacy naming"):
        units[0].initializeDynamicsSC(system.dynManager)


def test_deprecated_unit_rejects_renaming_after_preparation():
    """Changing a unit's prefix cannot leave existing state bindings under stale names."""
    system, units, panels, loads, ports = make_legacy_system()
    system.initializeDynamics()
    original = units[0].hub.nameOfHubPosition
    units[0].spacecraftName = "renamed"
    with pytest.raises(bskLogging.BasiliskError, match="rebuild the unit"):
        units[0].initializeDynamicsSC(system.dynManager)
    assert units[0].hub.nameOfHubPosition == original


if __name__ == "__main__":
    raise SystemExit(pytest.main([__file__]))

File: src/simulation/dynamics/stateArchitecture/_UnitTest/test_effectorNamingIntegration.cpp
/* Copyright (c) 2026, Autonomous Vehicle Systems Lab, University of Colorado at Boulder
 * This file is distributed under the ISC License in LICENSE.
 */

#include "architecture/_GeneralModuleFiles/sys_model.h"
#include "architecture/utilities/bskLogging.h"
#include "simulation/dynamics/HingedRigidBodies/hingedRigidBodyStateEffector.h"
#include "simulation/dynamics/_GeneralModuleFiles/dynParamManager.h"
#include "simulation/dynamics/_GeneralModuleFiles/dynamicEffector.h"
#include "simulation/dynamics/_GeneralModuleFiles/effectorName.h"
#include "simulation/dynamics/_GeneralModuleFiles/stateEffector.h"
#include "simulation/dynamics/prescribedMotion/prescribedMotionStateEffector.h"
#include "simulation/dynamics/spacecraft/spacecraft.h"
#include "simulation/dynamics/spinningBodies/spinningBodiesOneDOF/spinningBodyOneDOFStateEffector.h"
#include <Eigen/Dense>
#include <array>
#include <gtest/gtest.h>
#include <string>
#include <vector>

namespace {

class PreparationObserver : public DynamicEffector
{
  public:
    std::vector<std::string> requiredStates;
    std::vector<std::string> requiredProperties;
    std::array<Eigen::MatrixXd*, 4> properties{};
    int bindingCount = 0;

    void linkInStates(DynParamManager& manager) override { this->checkRegistration(manager); }
    void computeForceTorque(double, double) override {}
    void linkInProperties(DynParamManager& manager) override
    {
        this->checkRegistration(manager);
        this->properties = { manager.getPropertyReference(this->propName_inertialPosition),
                             manager.getPropertyReference(this->propName_inertialVelocity),
                             manager.getPropertyReference(this->propName_inertialAttitude),
                             manager.getPropertyReference(this->propName_inertialAngVelocity) };
        ++this->bindingCount;
    }

  private:
    void checkRegistration(const DynParamManager& manager) const
    {
        for (const auto& name : this->requiredStates) {
            EXPECT_EQ(manager.stateContainer.stateMap.count(name), 1U) << name;
        }
        for (const auto& name : this->requiredProperties) {
            EXPECT_EQ(manager.dynProperties.count(name), 1U) << name;
        }
    }
};

class ExternalEffector : public StateEffector
{
  public:
    int registrationCount = 0;
    void registerStates(DynParamManager&) override { ++this->registrationCount; }
    void linkInStates(DynParamManager&) override {}
    void computeDerivatives(double, Eigen::Vector3d, Eigen::Vector3d, Eigen::MRPd) override {}
};

class TaggedExternalEffector
  : public ExternalEffector
  , public SysModel
{};

void
configure(Spacecraft& vehicle)
{
    vehicle.hub.mHub = 100.0;                                      // [kg]
    vehicle.hub.IHubPntBc_B = 100.0 * Eigen::Matrix3d::Identity(); // [kg m^2]
    vehicle.dynManager.setEffectorNamingPolicy(EffectorNamingPolicy::ManagerLocal);
}

struct Branches
{
    // Construct the last attached leaf first, so constructor counters cannot determine order.
    SpinningBodyOneDOFStateEffector last;
    SpinningBodyOneDOFStateEffector first;
    PrescribedMotionStateEffector secondRoot;
    PrescribedMotionStateEffector firstRoot;
    PreparationObserver firstObserver;
    PreparationObserver lastObserver;

    Branches()
    {
        first.sHat_S = Eigen::Vector3d::UnitX();
        last.sHat_S = Eigen::Vector3d::UnitY();
        first.thetaInit = 0.125; // [rad]
        last.thetaInit = -0.25;  // [rad]
        // A later custom declaration must reserve the earlier automatic candidate.
        last.setNameOfThetaState("prescribedObjectspinningBodyTheta1");
        last.setNameOfInertialPositionProperty("spinningBodyInertialPosition1");
        firstRoot.setNameOfsigma_PMState("firstFrame");
        secondRoot.setNameOfsigma_PMState("lastFrame");
        firstRoot.addStateEffector(&first);
        secondRoot.addStateEffector(&last);
        first.addDynamicEffector(&firstObserver);
        last.addDynamicEffector(&lastObserver);
        for (auto* observer : { &firstObserver, &lastObserver }) {
            observer->requiredStates = {
                "firstFrame", "lastFrame", "prescribedObjectspinningBodyTheta2", "prescribedObjectspinningBodyTheta1"
            };
            observer->requiredProperties = { "spinningBodyInertialPosition2", "spinningBodyInertialPosition1" };
        }
    }

    std::vector<StateEffector*> roots() { return { &firstRoot, &secondRoot }; }
};

TEST(EffectorNamingIntegration, spacecraftRegistersEveryBranchBeforeBinding)
{
    Branches branches;
    Spacecraft vehicle;
    configure(vehicle);
    for (auto* root : branches.roots()) {
        vehicle.addStateEffector(root);
    }
    PreparationObserver hubObserver;
    hubObserver.requiredStates = branches.firstObserver.requiredStates;
    hubObserver.requiredProperties = branches.firstObserver.requiredProperties;
    vehicle.addDynamicEffector(&hubObserver);
    vehicle.initializeDynamics();
    EXPECT_EQ(branches.firstObserver.bindingCount, 1);
    EXPECT_EQ(branches.lastObserver.bindingCount, 1);
    EXPECT_EQ(branches.first.getNameOfThetaState(), "prescribedObjectspinningBodyTheta2");
    EXPECT_EQ(branches.last.getNameOfThetaState(), "prescribedObjectspinningBodyTheta1");
    EXPECT_NE(branches.firstObserver.properties[0], branches.lastObserver.properties[0]);
    const auto firstProperties = branches.firstObserver.properties;
    const auto lastProperties = branches.lastObserver.properties;
    auto* firstState = vehicle.dynManager.getStateObject(branches.first.getNameOfThetaState());
    const auto stateCount = vehicle.dynManager.stateContainer.stateMap.size();
    const auto propertyCount = vehicle.dynManager.dynProperties.size();
    vehicle.initializeDynamics();
    EXPECT_EQ(branches.firstObserver.bindingCount, 2);
    EXPECT_EQ(branches.lastObserver.bindingCount, 2);
    EXPECT_EQ(branches.firstObserver.properties, firstProperties);
    EXPECT_EQ(branches.lastObserver.properties, lastProperties);
    EXPECT_EQ(vehicle.dynManager.getStateObject(branches.first.getNameOfThetaState()), firstState);
    EXPECT_EQ(vehicle.dynManager.stateContainer.stateMap.size(), stateCount);
    EXPECT_EQ(vehicle.dynManager.dynProperties.size(), propertyCount);
    EXPECT_DOUBLE_EQ(firstState->getState()(0, 0), branches.first.thetaInit);
}

TEST(EffectorNamingIntegration, sharedManagerPreparesMultipleTreesAndAllTheirDescendants)
{
    Branches branches;
    PrescribedMotionStateEffector intermediate;
    SpinningBodyOneDOFStateEffector deepLeaf;
    deepLeaf.sHat_S = Eigen::Vector3d::UnitZ();
    deepLeaf.setNameOfThetaState("deepAngle");
    intermediate.setNameOfsigma_PMState("middleFrame");
    intermediate.addStateEffector(&deepLeaf);
    branches.firstRoot.addStateEffector(&intermediate);
    PreparationObserver deepObserver;
    deepLeaf.addDynamicEffector(&deepObserver);
    for (auto* observer : { &branches.firstObserver, &branches.lastObserver, &deepObserver }) {
        observer->requiredStates = {
            "firstFrame", "middleFrame", "deepAngle", "lastFrame", "prescribedObjectspinningBodyTheta1"
        };
        observer->requiredProperties = { "spinningBodyInertialPosition1" };
    }
    DynParamManager manager;
    manager.setEffectorNamingPolicy(EffectorNamingPolicy::ManagerLocal);
    const auto roots = branches.roots();
    StateEffector::collectEffectorNames(manager, roots);
    // Fixed names from all owners must participate before resolving the shared manager.
    manager.registerState(1, 1, "prescribedObjectspinningBodyTheta2");
    manager.createProperty("spinningBodyInertialPosition3", Eigen::Vector3d::Zero());
    manager.resolveEffectorNames();
    // This checks naming preparation at arbitrary depth, without assuming support for
    // the dynamics of a prescribed frame attached to another prescribed frame.
    for (auto* root : roots) {
        root->registerStates(manager);
    }
    for (auto* root : roots) {
        root->bindAttachedDynamicEffectors(manager);
    }
    EXPECT_EQ(branches.first.getNameOfThetaState(), "prescribedObjectspinningBodyTheta4");
    EXPECT_EQ(branches.last.getNameOfThetaState(), "prescribedObjectspinningBodyTheta1");
    EXPECT_EQ(deepObserver.bindingCount, 1);
    EXPECT_NE(deepObserver.properties[0], branches.firstObserver.properties[0]);
    const auto stateCount = manager.stateContainer.stateMap.size();
    const auto propertyCount = manager.dynProperties.size();
    const auto properties = deepObserver.properties;
    StateEffector::collectEffectorNames(manager, roots);
    manager.resolveEffectorNames();
    for (auto* root : roots) {
        root->registerStates(manager);
    }
    for (auto* root : roots) {
        root->bindAttachedDynamicEffectors(manager);
    }
    EXPECT_EQ(deepObserver.bindingCount, 2);
    EXPECT_EQ(deepObserver.properties, properties);
    EXPECT_EQ(manager.stateContainer.stateMap.size(), stateCount);
    EXPECT_EQ(manager.dynProperties.size(), propertyCount);
}

TEST(EffectorNamingIntegration, unsupportedExternalEffectorFailsBeforeAnyRegistrationOrBinding)
{
    for (const bool nested : { false, true }) {
        SCOPED_TRACE(nested);
        Spacecraft vehicle;
        configure(vehicle);
        HingedRigidBodyStateEffector supported;
        PreparationObserver observer;
        supported.addDynamicEffector(&observer);
        vehicle.addStateEffector(&supported);
        PrescribedMotionStateEffector root;
        PrescribedMotionStateEffector branch;
        TaggedExternalEffector external;
        external.ModelTag = "externalPanel%name";
        if (nested) {
            root.addStateEffector(&branch);
            branch.addStateEffector(&external);
            vehicle.addStateEffector(&root);
        } else {
            vehicle.addStateEffector(&external);
        }
        try {
            vehicle.initializeDynamics();
            FAIL() << "Unsupported external effector was accepted";
        } catch (const BasiliskError& error) {
            const std::string message = error.what();
            EXPECT_NE(message.find(external.ModelTag), std::string::npos);
            EXPECT_NE(message.find("describeEffectorNames()"), std::string::npos);
            EXPECT_NE(message.find("legacy naming"), std::string::npos);
        }
        EXPECT_TRUE(vehicle.dynManager.stateContainer.stateMap.empty());
        EXPECT_TRUE(vehicle.dynManager.dynProperties.empty());
        EXPECT_EQ(external.registrationCount, 0);
        EXPECT_EQ(observer.bindingCount, 0);
    }
}

TEST(EffectorNamingIntegration, legacyStillAcceptsExternalEffectors)
{
    Spacecraft vehicle;
    vehicle.hub.mHub = 100.0; // [kg]
    ExternalEffector external;
    vehicle.addStateEffector(&external);
    EXPECT_NO_THROW(vehicle.initializeDynamics());
    EXPECT_EQ(external.registrationCount, 1);
}

} // namespace
